### PR TITLE
Make non-matching interpolation more robust and deterministic

### PR DESCRIPTION
**Pull request by jorgensd**
**Current approach**
1. Create process bounding box tree with padding (`bb_global`)
2. For all local interpolation points find intersection with `bb_global`
3. Send these points to those processes
4. Find cells on process whose bounding box intersect with these points
5. Find actual cell collisions ( `compute_first_colliding_cell` that has a fixed tolerance. This uses GJK that has another fixed tolerance).
6. Send these back to process owning points. This process chooses first process that has a collision (non-deterministic).
7. Send owner info to all relevant process to reduce the processes searching later on.

As we see in this approach step 5 is problematic, as they require tolerances (that currently aren't in sync, causing #2774 ), which also are tricky to tune when we get to variable precision of geometry.
Step 6 is also problematic as it depends on the MPI neighborhood comm layout.

**New approach**

1. Create process bounding box tree with padding (`bb_global`)
2. For all local interpolation points find intersection with `bb_global`
3. Send these points to those processes
4. Find cells on process whose bounding box intersect with these points
5. Find closest entity on process.
  a. Uses midpoint trees to get a good starting point in bounding box tree
  b. Finds closest entity within GJK distance (single tolerance)
7. Send these back to process owning points. This process chooses the entity with the smallest distance (this means that for points far outside the mesh we use extrapolation).
8. Send owner info to all relevant process to reduce the processes searching later on.

Currently the extrapolation is limited by the padding we give to the bounding box in
`determine_point_ownership`, but this could now trivially be extended as an input parameter, both to `determine_point_ownership` and `create_nonmatching_meshes_interpolation_data`.

### CODE REVIEW

**Comment by jorgensd** [comment]
@massimiliano-leoni I would like you to review this as well, but cant add you as reviewer on github!

File: cpp/dolfinx/geometry/utils.h [hunk 1]
```diff
@@ -791,20 +786,30 @@ determine_point_ownership(const mesh::Mesh<T>& mesh, std::span<const T> points)
...
-      recv_sizes.data(), recv_offsets.data(),
-      dolfinx::MPI::mpi_type<std::int32_t>(), reverse_comm);
+      squared_distances.data(), send_sizes.data(), send_offsets.data(),
+      dolfinx::MPI::mpi_type<T>(), recv_distances.data(), recv_sizes.data(),
+      recv_offsets.data(), dolfinx::MPI::mpi_type<T>(), reverse_comm);
 
   std::vector<std::int32_t> point_owners(points.size() / 3, -1);
-  for (std::size_t i = 0; i < unpack_map.size(); i++)
+  std::vector<T> closest_distance(points.size() / 3, -1);
+  std::stringstream qq;
```
Comment (massimiliano-leoni) [inline cpp/dolfinx/geometry/utils.h:798]: ?

Comment (jorgensd) [reply cpp/dolfinx/geometry/utils.h:798]: Stringstream is remaining from debugging. Will remove

Comment (jorgensd) [inline cpp/dolfinx/geometry/utils.h:798]:
```suggestion
```

File: cpp/dolfinx/geometry/utils.h [hunk 2]
```diff
@@ -791,20 +786,29 @@ determine_point_ownership(const mesh::Mesh<T>& mesh, std::span<const T> points)
...
-    // Only insert new owner if no owner has previously been found
-    if ((recv_ranks[i] >= 0) && (point_owners[pos] == -1))
-      point_owners[pos] = recv_ranks[i];
+    for (std::int32_t j = recv_offsets[i]; j < recv_offsets[i + 1]; j++)
+    {
+      const std::int32_t pos = unpack_map[j];
+      // If point has not been found yet distance is negative
+      // If new received distance smaller than current distance choose owner
+      if ((closest_distance[pos] < 0)
+          || (closest_distance[pos] > recv_distances[j]))
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/utils.h:806]:
```suggestion
      if ( auto d = closest_distance[pos]; d < 0 or d > recv_distances[j])
```

File: cpp/dolfinx/geometry/utils.h [hunk 3]
```diff
@@ -753,18 +754,12 @@ determine_point_ownership(const mesh::Mesh<T>& mesh, std::span<const T> points)
...
-  {
-    std::array<T, 3> point;
-    std::copy(std::next(received_points.begin(), p),
-              std::next(received_points.begin(), p + 3), point.begin());
-    const int colliding_cell
-        = geometry::compute_first_colliding_cell(mesh, bb, point);
-    cell_indicator[p / 3] = (colliding_cell >= 0) ? rank : -1;
-    colliding_cells[p / 3] = colliding_cell;
-  }
+  std::vector<std::int32_t> closest_cells = compute_closest_entity(
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/utils.h:757]: Should this be `const`?

File: cpp/dolfinx/geometry/utils.h [hunk 4]
```diff
@@ -753,18 +754,12 @@ determine_point_ownership(const mesh::Mesh<T>& mesh, std::span<const T> points)
...
+      bb, midpoint_tree, mesh,
+      std::span<const T>(received_points.data(), received_points.size()));
+  std::vector<T> squared_distances = squared_distance(
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/utils.h:760]: `const`?

File: cpp/dolfinx/geometry/utils.h [hunk 5]
```diff
@@ -753,18 +754,12 @@ determine_point_ownership(const mesh::Mesh<T>& mesh, std::span<const T> points)
 
   // Each process checks which points collides with a cell on the process
```
Comment (IgorBaratta) [inline cpp/dolfinx/geometry/utils.h:755]: Do we need to change this comment?

**Comment by IgorBaratta** [comment]
Do we need to change the documentation of the function?
It seems that the PR is changing its behavior.

File: cpp/dolfinx/geometry/utils.h (hunk 5 above)
Comment (jorgensd) [inline cpp/dolfinx/geometry/utils.h:755]:
```suggestion
  // Each process checks which local cell is closest and computes the squared distance to the cell
```

**Comment by jorgensd** [comment]
Any further comments?

**Comment by garth-wells** [comment]
> Any further comments?

Just hoping we can get the oneAPI CI working. It's LLVM-based, whereas other CI is gcc.

**Pull request merged**